src/Sylius/Bundle/CoreBundle/Twig/ChannelProductUrlExtension.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Twig;
+
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Core\Model\ProductTranslation;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Symfony\Component\Routing\RouterInterface;
+use Twig\Extension\AbstractExtension;
+use Twig\TwigFunction;
+
+final class ChannelProductUrlExtension extends AbstractExtension
+{
+    public function __construct(
+        private ChannelUrlExtension $channelUrlExtension,
+        private LocaleContextInterface $localeContext,
+        private RouterInterface $router,
+    ) {
+    }
+
+    public function getFunctions(): array
+    {
+        return [
+            new TwigFunction('sylius_channel_url_for_product', [$this, 'generateChannelUrlForProduct']),
+        ];
+    }
+
+    public function generateChannelUrlForProduct(ProductInterface $product, ChannelInterface $channel): string
```

Comment:
This should return null when it's not possible to generate an URL.
It's a very unlikely scenario, but still possible.

Reply:
Nope this should not be possible if you correctly create the product. You can not create a product without setting at least one slug. The data structure would allow that but the workflow around it shouldn't.